src/main/MQ2Utilities.cpp

```diff
@@ -1310,58 +1310,31 @@ int GetLanguageIDByName(const char* szName)
-	if (!_stricmp(szName, "Fists of Bayle")) return ALTCURRENCY_FISTSOFBAYLE;
-	if (!_stricmp(szName, "Nobles")) return ALTCURRENCY_NOBLES;
-	if (!_stricmp(szName, "Arx Energy Crystals")) return ALTCURRENCY_ENERGYCRYSTALS;
-	if (!_stricmp(szName, "Pieces of Eight")) return ALTCURRENCY_PIECESOFEIGHT;
-	if (!_stricmp(szName, "Remnants of Tranquility")) return ALTCURRENCY_REMNANTSOFTRANQUILITY;
-	if (!_stricmp(szName, "Bifurcated Coin")) return ALTCURRENCY_BIFURCATEDCOIN;
-	if (!_stricmp(szName, "Adoptive Coins")) return ALTCURRENCY_ADOPTIVE;
-	if (!_stricmp(szName, "Sathir's Trade Gems")) return ALTCURRENCY_SATHIRSTRADEGEMS;
-	if (!_stricmp(szName, "Ancient Sebilisian Coins")) return ALTCURRENCY_ANCIENTSEBILISIANCOINS;
-	if (!_stricmp(szName, "Bathezid Trade Gems")) return ALTCURRENCY_BATHEZIDTRADEGEMS;
-	if (!_stricmp(szName, "Ancient Draconic Coin")) return ALTCURRENCY_ANCIENTDRACONICCOIN;
-	if (!_stricmp(szName, "Fetterred Ifrit Coins")) return ALTCURRENCY_FETTERREDIFRITCOINS;
-	if (!_stricmp(szName, "Entwined Djinn Coins")) return ALTCURRENCY_ENTWINEDDJINNCOINS;
-	if (!_stricmp(szName, "Crystallized Luck")) return ALTCURRENCY_CRYSTALLIZEDLUCK;
-	if (!_stricmp(szName, "Froststone Ducat")) return ALTCURRENCY_FROSTSTONEDUCAT;
-	if (!_stricmp(szName, "Warlord's Symbol")) return ALTCURRENCY_WARLORDSSYMBOL;
-	if (!_stricmp(szName, "Overseer Tetradrachm")) return ALTCURRENCY_OVERSEERTETRADRACHM;
-	if (!_stricmp(szName, "Restless Mark")) return ALTCURRENCY_RESTLESSMARK;
-	if (!_stricmp(szName, "Warforged Emblem")) return ALTCURRENCY_WARFORGEDEMBLEM;
-	if (!_stricmp(szName, "Scarlet Marks")) return ALTCURRENCY_SCARLETMARKS;
-	if (!_stricmp(szName, "Medals of Conflict")) return ALTCURRENCY_MEDALSOFCONFLICT;
-	if (!_stricmp(szName, "Shaded Specie")) return ALTCURRENCY_SHADEDSPECIE;
-	if (!_stricmp(szName, "Spiritual Medallion")) return ALTCURRENCY_SPIRITUALMEDALLION;
-	if (!_stricmp(szName, "Laurion Inn Voucher")) return ALTCURRENCY_LAURIONINNVOUCHER;
-	if (!_stricmp(szName, "Shalowains Private Reserve")) return ALTCURRENCY_SHALOWAINSPRIVATERESERVE;
+int GetCurrencyIDByName(const char* szName)
+{
+	constexpr std::string_view chars_to_remove = "'`";
+	const std::string currency = remove_chars(szName, chars_to_remove);
+	for (int i = ALTCURRENCY_FIRST; i <= ALTCURRENCY_LAST; ++i)
+	{
+		// Check the plural form first
+		if (const char* currency_name_plural = pCDBStr->GetString(i, eAltCurrencyNamePlural))
+		{
+			if (ci_equals(currency, remove_chars(currency_name_plural, chars_to_remove)))
+				return i;
+
+			// Then check the singular form
+			if (const char* currency_name_singular = pCDBStr->GetString(i, eAltCurrencyName))
+			{
+				if (ci_equals(currency, remove_chars(currency_name_singular, chars_to_remove)))
+					return i;
+			}
+		}
+	}
+
+	// Crowns sit outside the ALTCURRENCY_LAST range
+	if (ci_equals(currency, "Crowns") || ci_equals(currency, "Crown"))
+		return ALTCURRENCY_CROWNS;
```

Comment:
Given the amount of work this function is doing, I would expect that we build a lookup table first.

Reply:
Implemented a cache as a gut check.  Concerns would be that there's no eviction right now.  I could go down the road of implementing a cache class that tracks LFU...buuuut...

Reply:
This doesn't really solve my concern. All the known values to compare about are known and do not change, so I'd prefer we build a lookup table and then only ever use that.

With this implementation, we will always call remove_chars on the input, and remove_chars on every string that we didn't already match, when that work is redundant.

just build the lookup table, and check if parameter is in the lookup table.

Reply:
Something like this? 

```cpp
void UpdateCurrencyCache(std::unordered_map<std::string, int>& cache, int value, eDatabaseStringType type)
{
	constexpr std::string_view chars_to_remove = "'`";
	if (const char* ptr = pCDBStr->GetString(value, type))
	{
		const std::string currency = to_lower_copy(ptr);
		cache[currency] = value;
		cache[remove_chars(currency, chars_to_remove)] = value;
	}
}

int GetCurrencyIDByName(const char* szName)
{
	static std::unordered_map<std::string, int> cache;

	if (cache.empty())
	{
		for (int i = ALTCURRENCY_FIRST; i <= ALTCURRENCY_LAST; ++i)
		{
			UpdateCurrencyCache(cache, i, eAltCurrencyNamePlural);
			UpdateCurrencyCache(cache, i, eAltCurrencyName);
		}
		// Crowns are outside ALTCURRENCY_LAST
		UpdateCurrencyCache(cache, ALTCURRENCY_CROWNS, eAltCurrencyNamePlural);
		UpdateCurrencyCache(cache, ALTCURRENCY_CROWNS, eAltCurrencyName);
	}

	const auto it = cache.find(to_lower_copy(szName));
	if (it != cache.end())
	{
		return it->second;
	}

	return -1;
}
```

Reply:
yep

Reply:
Updated